### ISSUE TITLE
`torch.musa._initialized` is a stale re-export, silently disabling checkpoint RNG preservation

### ISSUE THREAD

**Issue by yinjiew** [MooreThreads/torch_musa#150]
I don't have permission to add labels here — this is a correctness bug (silent wrong gradients), suggest `bug` / `API`. A fix + regression tests are ready and I'll open a PR against `main` referencing this issue.

## Affected versions

Every released `torch_musa` I could inspect is affected, and the defect is unchanged on
`main`:

| `torch_musa` | `torch_musa/__init__.py` re-exports `_initialized` | required PyTorch | `checkpoint.py` reads `getattr(device_module, "_initialized", False)` | affected |
|---|---|---|---|---|
| 2.7.1 | yes, line 112 | 2.7.1 | yes, 3 sites | **yes** |
| 2.9.1.post1 | yes, line 120 | 2.9.1 | yes, 3 sites | **yes** |
| 2.11.0.post1 (`main`) | yes, line 120 | 2.11.0 | yes, 3 sites | **yes** |

`torch_musa/__init__.py` is byte-identical between 2.9.1.post1 and 2.11.0.post1 apart from
the four PyTorch version strings, and `torch_musa/core/_lazy_init.py` is byte-identical
across all three. The three PyTorch versions gate identically, in
`CheckpointFunction.forward` and twice in `_checkpoint_without_reentrant_generator`.
Nothing in `torch_patches/` touches `torch/utils/checkpoint.py` in any release, so the
consumer is stock PyTorch in every case.

## Summary

`torch_musa` does set `_initialized = True` on successful lazy initialization — but it sets
it in `torch_musa.core._lazy_init`, while `torch_musa/__init__.py` re-exports the flag with
`from ... import`. In Python that binds a **snapshot of the value**, not a reference to the
name, so `torch_musa._initialized` is frozen at `False` forever.

`torch.musa` *is* the `torch_musa` module (`torch/__init__` gets it via
`torch.__setattr__("musa", sys.modules[__name__])`, `torch_musa/__init__.py:57`), and
PyTorch reads this flag as `getattr(device_module, "_initialized", False)` — notably in
`torch.utils.checkpoint`, to decide whether to save and restore **device** RNG state.

The consequence: on MUSA, `torch.utils.checkpoint(..., preserve_rng_state=True)` silently
degrades to CPU-RNG-only. Any stochastic operator inside a checkpointed region draws
*different* random numbers in the backward recompute than it did in the forward pass, which
produces a **numerically wrong gradient with no error, no exception, and no warning**.

`torch_musa.is_initialized()` is a *function*, so it reads the live value and returns the
correct answer. That is why every internal consumer behaves correctly and only external code
reading the attribute directly — i.e. upstream PyTorch — is affected.

## Impact

Affects any model wrapping a stochastic operator in `torch.utils.checkpoint`, which is
routine for memory-bound training:

- `nn.Dropout` inside a checkpointed transformer block
- stochastic depth / DropPath on residual branches
- Gumbel-softmax or other reparameterized sampling
- any custom activation that samples noise

Measured on 2x MTT S5000, a custom stochastic activation (two truncated-normal draws plus a
Gumbel straight-through branch) at ResNet-18 scale:

| quantity | max abs gradient difference, checkpointed vs not |
|---|---|
| stochastic activation | **2.6** |
| same, with a learnable branch | **139.7** |
| minimal `randn_like` repro below | **~4.7** |

These are not roundoff-scale discrepancies — the recompute samples an entirely independent
noise tensor. Training still converges to *something*, which is exactly why this can go
unnoticed indefinitely.

## Environment

```text
torch        2.7.1a0+gite2d141d
torch_musa   2.7.1+5ee0a64     (measured on hardware)
torch_musa   2.9.1.post1       (source-reviewed; defect unchanged)
torch_musa   2.11.0.post1      (source-reviewed; defect unchanged; == branch main)
GPU          2 x Moore Threads MTT S5000 (80 GB)
Python       3.10
OS           Ubuntu (container)
```

### The stale re-export

`torch_musa/core/_lazy_init.py`:

```python
 19:  _initialized = False
...
 77:  def is_initialized():
 79:      return _initialized and not _is_in_bad_fork()
...
 95:  def _lazy_init():
100:      global _initialized, _queue_calls
...
137:          _initialized = True          # rebinds torch_musa.core._lazy_init._initialized
```

`torch_musa/__init__.py`:

```python
 57:  torch.__setattr__("musa", sys.modules[__name__])     # torch.musa IS torch_musa
...
117:  from .core._lazy_init import (
118:      _lazy_init,
119:      _lazy_call,
120:      _initialized,        # <-- binds a SNAPSHOT of False, taken at import time
121:      _is_in_bad_fork,
122:      is_initialized,
123:      musart,
124:  )
```

Line 120 copies the *value* `False` into the `torch_musa` namespace. When line 137 later
rebinds the name inside `torch_musa.core._lazy_init`, that rebinding cannot propagate to
`torch_musa.__dict__["_initialized"]` — they are two independent bindings that merely
happened to start out pointing at the same object.

`is_initialized` (line 122) is a *function object*; copying the binding is harmless because
calling it re-reads the module global each time. Hence the asymmetry: the function is right,
the attribute is wrong.

### Why `torch.cuda` does not have this bug

In `torch/cuda/__init__.py`, `_initialized` and `_lazy_init()` live in the **same module**,
so `global _initialized` rebinds exactly the attribute that `torch.cuda._initialized`
resolves to. `torch_musa` factored lazy-init out into a submodule and re-exported the flag,
which is what introduced the divergence.

### The consumer

`torch/utils/checkpoint.py`, in `_checkpoint_without_reentrant_generator`:

```python
if preserve_rng_state:
    fwd_cpu_state = torch.get_rng_state()
    had_device_in_fwd = False
    device_module = _get_device_module(device_type)
    if getattr(device_module, "_initialized", False):     # <-- always False on MUSA
        had_device_in_fwd = True
        fwd_devices, fwd_device_states = get_device_states(*args)
```

Note the `getattr(..., False)` default. "Device not initialized" is a legitimate state for
which skipping device-RNG capture is *correct* behaviour, so PyTorch rightly emits no
warning. That silent-and-correct path is simply being reached under a false premise.

The **reentrant** path is gated identically — `CheckpointFunction.forward` in the same
file:

```python
ctx.had_device_in_fwd = False
device_module = _get_device_module(ctx.device_type)
if getattr(device_module, "_initialized", False):     # <-- always False on MUSA
    ctx.had_device_in_fwd = True
    ctx.fwd_devices, ctx.fwd_device_states = get_device_states(*args)
```

so both `use_reentrant=True` and `use_reentrant=False` are affected. `checkpoint.py` reads
the flag in exactly three places; the third is the guard that is supposed to *reject* a
device initialized for the first time inside a checkpointed forward, which the stale value
also disarms:

```python
if getattr(device_module, "_initialized", False) and \
   preserve_rng_state and not had_device_in_fwd:
    raise RuntimeError(
        "PyTorch's device state was initialized in the forward pass "
        "of a Checkpoint, which is not allowed. ...")
```

On MUSA that `RuntimeError` can never fire either, so the one case PyTorch considers
unsupportable is also accepted silently.

Everything else in the chain already works, verified individually:

| link | status |
|---|---|
| `_infer_device_type(tensor)` returns `'musa'` | works |
| `_get_device_module('musa')` returns the `torch_musa` module | works |
| `torch.musa.get_rng_state()` / `set_rng_state()` round-trip bitwise | works |
| `torch.musa.device(idx)` context manager | works |
| `torch.musa.is_initialized()` | works (function, reads live value) |
| `torch.musa._initialized` | **frozen `False` — gates off all of the above** |

## Minimal demonstration of the mechanism

No MUSA hardware or PyTorch needed — this is pure Python semantics:

```python
# pkg/core/_lazy.py
_initialized = False
def is_initialized(): return _initialized
def _lazy_init():
    global _initialized
    _initialized = True

# pkg/__init__.py
from .core._lazy import _lazy_init, _initialized, is_initialized
```

```python
import pkg
from pkg.core import _lazy

pkg._lazy_init()
print(pkg._initialized)        # False  <-- re-exported snapshot, stale
print(_lazy._initialized)      # True   <-- authoritative
print(pkg.is_initialized())    # True   <-- function reads live value
```

A self-contained script that builds both the broken package and the proposed fix and
prints the comparison is attached below. It needs neither MUSA nor PyTorch, so it can be
checked on any machine in a few seconds:

```text
re-export with `from ... import _initialized`  (torch_musa today)
  broken  before init                     : False
          core._lazy_init._initialized     : True   <- authoritative
          is_initialized()                 : True   <- function, correct
          getattr(pkg,'_initialized',False): False   <- what checkpoint reads
          missing attribute still raises   : AttributeError (ok)

forward with a module-level `__getattr__`      (proposed fix)
  fixed   before init                     : False
          core._lazy_init._initialized     : True   <- authoritative
          is_initialized()                 : True   <- function, correct
          getattr(pkg,'_initialized',False): True   <- what checkpoint reads
          missing attribute still raises   : AttributeError (ok)

result:
  broken -> checkpoint would skip the device RNG : True
  fixed  -> checkpoint saves the device RNG      : True
```

## Reproduction on hardware

```python
import torch
import torch_musa
from torch.utils.checkpoint import checkpoint

torch.musa.set_device(0)
dev = torch.device("musa:0")
print("torch.musa is torch_musa     :", torch.musa is torch_musa)

# 1. the MUSA RNG accessors themselves are fine
state = torch.musa.get_rng_state()
a = torch.randn(1024, device=dev)
torch.musa.set_rng_state(state)
b = torch.randn(1024, device=dev)
print("RNG save/restore round-trips :", torch.equal(a, b))          # True

# 2. the flag disagrees with reality
from torch_musa.core import _lazy_init as _li
print("is_initialized()             :", torch_musa.is_initialized())  # True
print("_lazy_init._initialized      :", _li._initialized)             # True
print("torch_musa._initialized      :", torch_musa._initialized)      # False  <-- STALE

# 3. so checkpoint does not replay the noise
draws = []

def fn(t):
    n = torch.randn_like(t)              # any stochastic op inside the checkpoint
    draws.append(n.detach().clone())
    return t * n

x = torch.randn(256, device=dev, requires_grad=True)
checkpoint(fn, x, use_reentrant=False, preserve_rng_state=True).sum().backward()
print("passes observed              :", len(draws))                       # 2
print("forward noise == recompute   :", torch.equal(draws[0], draws[1]))  # False  <-- BUG
print("max abs difference           :", float((draws[0] - draws[1]).abs().max()))

# 4. un-staling the flag is sufficient
torch_musa._initialized = True
draws.clear()
x2 = torch.randn(256, device=dev, requires_grad=True)
checkpoint(fn, x2, use_reentrant=False, preserve_rng_state=True).sum().backward()
print("after refreshing the flag    :", torch.equal(draws[0], draws[1]))  # True
```

### Expected

`forward noise == recompute` is `True`. That is the entire contract of
`preserve_rng_state=True`, and it is what CUDA delivers.

### Actual

```text
torch.musa is torch_musa     : True
RNG save/restore round-trips : True
is_initialized()             : True
_lazy_init._initialized      : True
torch_musa._initialized      : False
passes observed              : 2
forward noise == recompute   : False
max abs difference           : 4.910468
after refreshing the flag    : True
```

## Why this went unnoticed

1. **Every internal caller uses `is_initialized()`**, which is correct. Only upstream
   PyTorch reads the raw attribute, so nothing inside `torch_musa` misbehaves.
2. **The existing checkpoint test cannot catch it.**
   `tests/unittest/amp/test_amp_checkpoint.py` checkpoints `nn.Linear` — fully
   deterministic — and asserts only `inputs.dtype == output.dtype`. It covers AMP dtype
   propagation through checkpoint, not the RNG contract. `test_amp_checkpoint_fp16` has no
   assertion at all.
3. **The one place that asserts the flag never executes.**
   `tests/unittest/miscs/test_musa_converter.py:70` contains `assert torch.musa._initialized`
   — but inside the `python_dst` string literal, as the *expected output* of the CUDA→MUSA
   source converter. It is compared as text, never run.

### Primary: stop re-exporting the flag; forward it with PEP 562 `__getattr__`

Keep `torch_musa.core._lazy_init` as the single source of truth and resolve the attribute
dynamically. Remove line 120:

```diff
--- a/torch_musa/__init__.py
+++ b/torch_musa/__init__.py
@@ -117,7 +117,6 @@
 from .core._lazy_init import (
     _lazy_init,
     _lazy_call,
-    _initialized,
     _is_in_bad_fork,
     is_initialized,
     musart,
 )
```

and add, at module level in `torch_musa/__init__.py`, immediately after that import block:

```python
from .core import _lazy_init as _lazy_init_mod


def __getattr__(name):
    """Resolve mutable lazy-init state against the module that owns it (PEP 562).

    PyTorch probes backend modules with `getattr(device_module, "_initialized", False)`.
    `torch.utils.checkpoint` uses that flag to decide whether to stash the device RNG
    state in the forward pass and replay it in the backward recompute, so a value stuck
    at `False` silently downgrades `preserve_rng_state=True` to CPU-RNG-only: every
    stochastic op inside a checkpointed region redraws, and the gradient is wrong with
    no error and no warning.
    """
    if name == "_initialized":
        return _lazy_init_mod._initialized
    raise AttributeError(f"module {__name__!r} has no attribute {name!r}")
```

`from .core import _lazy_init as _lazy_init_mod` binds the *module*, not the same-named
function, because `torch_musa/core/__init__.py` already aliases the function to
`_lazy_init_musa` for exactly this reason (`# avoid conflict with module _lazy_init`).
Hoisting the handle to module level keeps `__getattr__` a plain dict lookup plus one
attribute read.

`__getattr__` is only consulted when normal attribute lookup fails, so deleting line 120 is
required — otherwise the stale entry in `torch_musa.__dict__` keeps shadowing it.

This is correct for every path automatically, including re-initialization after fork and any
future error path that resets the flag, because there is only ever one binding.

A sibling of this technique is already in the tree: `torch_musa/core/mudnn.py` replaces its
own module object with a `PropModule` so that `torch.backends.mudnn.allow_tf32` reads
through to C++ rather than returning a stale Python copy. Same problem, same shape of
solution.

### Alternative, and why it is worse

Writing through from `_lazy_init()`:

```python
_initialized = True
torch_musa._initialized = True      # _lazy_init.py already imports torch_musa
```

This works and is a smaller diff, but creates two sources of truth that must be kept in sync
at every site that touches the flag — including bad-fork reset and any future teardown path.
The `__getattr__` form cannot drift.

### Note for either fix

`from torch_musa import _initialized` in *user* code still yields a snapshot. That is
inherent to Python's import semantics and cannot be fixed from inside the package; it is not
a concern for `torch.utils.checkpoint`, which uses `getattr` on the module object.

## Suggested regression tests

The valuable test is the end-to-end contract, not the flag. Two homes:

1. `tests/unittest/standalone/test_lazy_init.py` — the flag itself. This directory is the
   right place because `run_unittest.sh` launches each standalone file in its own process,
   which is what makes an assertion about lazy-init state meaningful.

   ```python
   def test_initialized_attr_is_live(self):
       x = torch.randn((8,), device="musa")
       assert torch_musa.core._lazy_init.is_initialized()
       assert torch_musa._initialized
       assert torch.musa._initialized
       del x
   ```

2. `tests/unittest/core/test_checkpoint.py` (new) — the contract, parametrized over both
   `use_reentrant` values, plus gradient equality against an uncheckpointed reference for
   `randn_like` and for `nn.Dropout`:

   ```python
   @pytest.mark.parametrize("use_reentrant", [False, True])
   @testing.test_on_nonzero_card_if_multiple_musa_device(1)
   def test_checkpoint_preserves_device_rng(use_reentrant):
       draws, grad = _noisy_mul_under_checkpoint(use_reentrant)
       assert len(draws) == 2, f"expected forward + 1 recompute, saw {len(draws)}"
       assert torch.equal(draws[0], draws[1]), "recompute drew different noise"
       assert torch.equal(grad, draws[0])       # d(x * n)/dx == n
   ```

Both tests fail on today's `main` and pass with the fix. A full patch containing the fix and
both tests is attached to the PR.

## Unrelated minor finding

`torch_musa/core/_lazy_init.py:100` declares `global _initialized, _queue_calls`, but the
module-level list is named `_queued_calls` (line 22). The declaration creates an unused
global; it is harmless today only because `_queued_calls` is mutated via `.append()` and
never rebound, so no `global` declaration is needed for it at all. Likely a typo worth
cleaning up.

## Scope of verification

- The stale binding and the wrong gradients were **measured** on hardware
  (2 x MTT S5000, `torch_musa` 2.7.1+5ee0a64) on the non-reentrant path
  (`use_reentrant=False`, the current PyTorch default).
- The pure-Python mechanism was reproduced in isolation (see the minimal demonstration
  above), so it does not depend on any MUSA component.
- The defect is **source-confirmed** in `torch_musa` 2.9.1.post1 and on branch `main`
  (2026-08-17): the re-export at `torch_musa/__init__.py:120` is unchanged.
- I no longer have access to the MTT S5000 machine, so I cannot re-run anything on
  2.11.0.post1 hardware. Everything above is either a measurement taken while I still had
  the machine (on 2.7.1), a source review of the released trees, or a pure-Python
  reproduction that anyone can re-run without a GPU. I have flagged which is which
  throughout, and I have not extrapolated a single number.
- The **reentrant** path is source-confirmed affected in PyTorch v2.9.1
  (`CheckpointFunction.forward` uses the same `getattr(device_module, "_initialized", False)`
  gate), but was not measured on hardware. The proposed tests cover both paths.

### One behavioural change to expect after the fix

The third read of the flag quoted above becomes live again: if MUSA is initialized for the
*first time* inside a checkpointed forward, `torch.utils.checkpoint` will now raise
`RuntimeError("PyTorch's device state was initialized in the forward pass of a
Checkpoint...")` instead of silently continuing. That is upstream's intended behaviour and
matches CUDA; it is hard to hit in practice (the checkpointed region would have to be the
first thing in the process to touch a MUSA tensor), but it is a real, deliberate difference
rather than a pure bug fix, so it belongs in the release note.
